### ISSUE TITLE
Video Capture is Empty when specifying BinaryFolder

### ISSUE THREAD

**Issue by nikeee** [rosenbjerg/FFMpegCore#256]
If I'm specyfing the path to the folder where the ffmpeg.exe resides, the captured video seems to be invalid and/or empty. The file has a non-zero-size, but VLC does not show anything.

When I'm using this, it does not work:
```csharp
await ffmpegInstance.ProcessAsynchronously(ffMpegOptions: new FFOptions()
{
    BinaryFolder = System.IO.Path.Combine(FFmpegManager.FFmpegAppDataPath, "bin"),
});
```
If I'm omitting the options, it works. That would be because I've also installed ffmpeg locally in my PATH (which is a different installation + version).

With that code above, ffmpeg starts and seems to work (it even draws the frame when capturing the screen). But as soon as I cancel it, the video file is corrupt/empty as mentioned above. The same code works with the version in PATH.

I tried multiple binaries. The ones from ffbinaries as well as LGPL and GPL versions from https://github.com/BtbN/FFmpeg-Builds/releases. Both versions (the one that works as well as the ones that don't work) are 64 bit on windows.

Any ideas?

**Comment by rosenbjerg** [comment]
Which different versions are you experiencing different behaviour between?

**Comment by nikeee** [comment]
The version in PATH (installed via chocolatey) that works:
```
C:\Users\nikeee>ffmpeg
ffmpeg version 4.4-essentials_build-www.gyan.dev Copyright (c) 2000-2021 the FFmpeg developers
  built with gcc 10.2.0 (Rev6, Built by MSYS2 project)
  configuration: --enable-gpl --enable-version3 --enable-static --disable-w32threads --disable-autodetect --enable-fontconfig --enable-iconv --enable-gnutls --enable-libxml2 --enable-gmp --enable-lzma --enable-zlib --enable-libsrt --enable-libssh --enable-libzmq --enable-avisynth --enable-sdl2 --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxvid --enable-libaom --enable-libopenjpeg --enable-libvpx --enable-libass --enable-libfreetype --enable-libfribidi --enable-libvidstab --enable-libvmaf --enable-libzimg --enable-amf --enable-cuda-llvm --enable-cuvid --enable-ffnvcodec --enable-nvdec --enable-nvenc --enable-d3d11va --enable-dxva2 --enable-libmfx --enable-libgme --enable-libopenmpt --enable-libopencore-amrwb --enable-libmp3lame --enable-libtheora --enable-libvo-amrwbenc --enable-libgsm --enable-libopencore-amrnb --enable-libopus --enable-libspeex --enable-libvorbis --enable-librubberband
  libavutil      56. 70.100 / 56. 70.100
  libavcodec     58.134.100 / 58.134.100
  libavformat    58. 76.100 / 58. 76.100
  libavdevice    58. 13.100 / 58. 13.100
  libavfilter     7.110.100 /  7.110.100
  libswscale      5.  9.100 /  5.  9.100
  libswresample   3.  9.100 /  3.  9.100
  libpostproc    55.  9.100 / 55.  9.100
```

One that did not work:
```
ffmpeg>ffmpeg_.exe
ffmpeg version 4.2.1 Copyright (c) 2000-2019 the FFmpeg developers
  built with gcc 9.1.1 (GCC) 20190807
  configuration: --enable-gpl --enable-version3 --enable-sdl2 --enable-fontconfig --enable-gnutls --enable-iconv --enabl
e-libass --enable-libdav1d --enable-libbluray --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enab
le-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-libshine --enable-libsnappy --enable-libsoxr --enabl
e-libtheora --enable-libtwolame --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 -
-enable-libxml2 --enable-libzimg --enable-lzma --enable-zlib --enable-gmp --enable-libvidstab --enable-libvorbis --enabl
e-libvo-amrwbenc --enable-libmysofa --enable-libspeex --enable-libxvid --enable-libaom --enable-libmfx --enable-amf --en
able-ffnvcodec --enable-cuvid --enable-d3d11va --enable-nvenc --enable-nvdec --enable-dxva2 --enable-avisynth --enable-l
ibopenmpt
  libavutil      56. 31.100 / 56. 31.100
  libavcodec     58. 54.100 / 58. 54.100
  libavformat    58. 29.100 / 58. 29.100
  libavdevice    58.  8.100 / 58.  8.100
  libavfilter     7. 57.100 /  7. 57.100
  libswscale      5.  5.100 /  5.  5.100
  libswresample   3.  5.100 /  3.  5.100
  libpostproc    55.  5.100 / 55.  5.100
```

I also took the exact binary of the PATH version, moved it somewhere else and used the containing folder as a `BinaryFolder`. That also did not work. So I think the problem does not involve a specific version of ffmpeg. Instead, I think it has to do something with the different configuration...?
I also tried setting the temp directory to some other values, which did not help.

**Comment by nikeee** [comment]
Just found out that it works independently of the version that's used.

If I prepend the `bin` Folder or the version I want to use to `PATH` before invoking ffmpeg, it works regardless of the version in use.

**Comment by nikeee** [comment]
I'm able to reproduce this issue on a separate machine.

Observations:
The binary that's started when using a Chocolatey version is not the actual ffmpeg binary. It's a wrapper binary ("shim") that starts the actual ffmpeg binary. **It has the path to the actual ffmpeg binary hard-coded**. That target binary can be exchanged for different versions and it works with every binary tried.

Plugging the shim into dnSpy, it basically does this:
```csharp
ProcessStartInfo processStartInfo = new ProcessStartInfo(process, arguments)
{
    UseShellExecute = false,
    WorkingDirectory = working_directory,
    RedirectStandardError = false,
    RedirectStandardOutput = false,
    RedirectStandardInput = false,
    CreateNoWindow = false,
    WindowStyle = ProcessWindowStyle.Minimized
};
CommandExecutor.RunningProcess.Start();
CommandExecutor.RunningProcess.WaitForExit();
```

What I'm suspecting here is one of the following:
- Because the `ProcessStartInfo` differ from these in `Instances`, there might be an issue
- `WaitForExit` actually lets ffmpeg exit gracefully somehow

I also checked the actual command that is executed. The first one works, the second one doesn't:

```
"C:\ProgramData\chocolatey\bin\ffmpeg.exe" -f gdigrab -r 30 -offset_x 100 -offset_y 100 -video_size 100x100 -show_region 0 -draw_mouse 1 -i "desktop" -movflags faststart -c:v libx264 -pix_fmt yuv420p "C:\output\lul2.mp4" -y

"C:\ProgramData\chocolatey\lib\ffmpeg\tools\ffmpeg\bin\ffmpeg.exe" -f gdigrab -r 30 -offset_x 100 -offset_y 100 -video_size 100x100 -show_region 0 -draw_mouse 1 -i "desktop" -movflags faststart -c:v libx264 -pix_fmt yuv420p "C:\output\lul2.mp4" -y
```

`C:\ProgramData\chocolatey\bin\ffmpeg.exe` calls `C:\ProgramData\chocolatey\lib\ffmpeg\tools\ffmpeg\bin\ffmpeg.exe` behind the scenes, as described above. So it has to do _something_ with that shim.

Edit:
Because I don't need the advanced capabilities of this library, I ended up writing my own FFMpeg wrapper class. In case someone needs it (AGPL-3.0): https://github.com/nikeee/HolzShots/commit/8b92ca805c10e650b7dbe93d977f5b27895b5ba8

**Comment by rosenbjerg** [comment]
The problem seems to related to the chocolatey shim so I'm closing this issue

**Comment by nikeee** [comment]
I'm fine with closing this ticket, but I just want to make sure that my previous post was not misunderstood in case someone runs into a similar issue:

It **works with** the chocolatey shim. It **does not work** with any other binary. It seems like the chocolatey shim is not the issue, but *a solution*.

**Comment by rosenbjerg** [comment]
Thank you for clarifying what you meant. 
I am developing (and using) this library on windows with ffmpeg and ffprobe binaries from ffbinaries (as linked in the readme) with none of these issue. I simply download the binaries for my platform and drop them in a folder that is in PATH. So the chocolatey shim definitely isn't needed.
Having looked a bit at why chocolatey encourages shims, I think the problem could be related to path. Could you try with absolute paths?
Since the shim "Runs the target executable where it is, which means all dependencies and other things used are all in the original location" (from https://docs.chocolatey.org/en-us/features/shim)

**Comment by nikeee** [comment]
> Having looked a bit at why chocolatey encourages shims, I think the problem could be related to path. Could you try with absolute paths?

I've posted the _exact_ command lines above, in case anyone wants to try it out. Note that if you change the binary that the shim points to, it also works. I think it has to do with something that the shim does. The command lines are the exact same, except the absolute path. When the binary path is not a shim, it does not work.

Note that I did not check any other operations except screen grabbing. Maybe plain conversions _do_ work, I didn't check that.
